<p>
  <a class="topmenu" href="#paper"><span class="nav">article</span></a>
  <a class="topmenu" href="#data"><span class="nav">data</span></a>
  <a class="topmenu" href="#code"><span class="nav">code</span></a>
  <a class="topmenu" href="#install"><span class="nav">install</span></a>
  <a class="topmenu" href="#tutorial"><span class="nav">tutorials</span></a>
  <a class="topmenu" href="#docs"><span class="nav">documentation</span></a>
  <a class="topmenu" href="#help"><span class="nav">help</span></a>
  <a class="topmenu" href="#contact"><span class="nav">contact</span></a>
</p>

<div id="top-left">
    <p>Welcome to the home of OmniPath, a comprehensive collection of
    molecular biology prior knowledge from 103 databases, with focus on
    literature curated human and rodent signaling pathways.</p>
    <p>
    OmniPath is available via a
    <a href="https://github.com/saezlab/pypath/blob/master/webservice.rst"
        target="_blank">
        web service,
    </a>
    an
    <a href="https://github.com/saezlab/OmnipathR" target="_blank">
        R package,
    </a>
    a
    <a href="https://apps.cytoscape.org/apps/omnipath" target="_blank">
        Cytoscape plug-in
    </a>
    <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz968/5689817?guestAccessKey=c47f593a-0017-43c9-b980-962003c5c318" target="_blank">
            (paper)
    </a>
    and a
    <a href="https://github.com/saezlab/pypath" target="_blank">
        Python module.
    </a>
    </p>
    <p>
    We also provide a large variety of software to analyze and integrate this
    prior knowledge.
    The <a href="https://github.com/saezlab/pypath" target="_blank">pypath</a>
    Python module builds the database segments in OmniPath and provides the
    most versatile API supported by many stand alone utilities.
    Saez Lab offers a software suite for functional omics data analysis and
    mechanistic modeling of intra- and inter-cellular signaling, find more
    details at <a href="https://saezlab.org/?#tool" starget="_blank">
    saezlab.org</a>.
    </p>
    <p>
        <strong>New: combined metabolic and signaling network in MetaPKN</strong>
        – By Aurelien Dugourd. Available at
        <a href="https://metapkn.omnipathdb.org/" target="_blank">
            https://metapkn.omnipathdb.org/
        </a>.
    </p>
    <p>
        <strong>New: our paper about the OmniPath Cytoscape app</strong>
        –
        <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz968/5689817?guestAccessKey=c47f593a-0017-43c9-b980-962003c5c318" target="_blank">
            F Ceccarelli, D Turei, A Gabor, J Saez-Rodriguez (2020)
                Bringing Data From Curated Pathway Resources to
                Cytoscape With OmniPath. <em>Bioinformatics</em> 36 (1).
        </a>
    </p>
    <p>
        <img
        src="https://static.omnipathdb.org/omnipath_cover_webpage_980px.jpeg"
        title="Illustration by Spencer Phillips, EMBL-EBI. This cover artwork
            is the result of a collaboration between Spencer Phillips,
            Mary Todd-Bergman, Denes Turei, Tamas Korcsmaros and
            Julio Saez-Rodriguez, starting with the initial idea of using
            lenses by Jakob Wirbel (RWTH-Aachen).
            A modified version of this figure appeared in the Cover of
            Nature Methods Dec 2016."
        style="width: 100%; max-width: 980px;"
        />
    </p>

    <p style="font-size: small; max-width: 980px;"><em>
    Illustration by Spencer Phillips, EMBL-EBI.
    This cover artwork is the result of a collaboration between
    Spencer Phillips, Mary Todd-Bergman, Denes Turei, Tamas Korcsmaros and
    Julio Saez-Rodriguez, starting with the initial idea of using lenses
    by Jakob Wirbel (RWTH-Aachen). A modified version of this figure appeared
    in the
    <a target="_blank"
        href="https://www.nature.com/nmeth/journal/v13/n12/covers/index.html">
        Cover of Nature Methods Dec 2016
    </a>
    </em></p>
    
</div>

<div id="top-right">
    <a
        class="twitter-timeline"
        data-lang="en"
        data-width="240"
        data-height="580"
        data-link-color="#FAB81E"
        href="https://twitter.com/omnipathdb?ref_src=twsrc%5Etfw">
        Tweets by omnipathdb
    </a>
    <script async src="https://platform.twitter.com/widgets.js"
        charset="utf-8">
    </script>
</div>

<div style="clear: both;"></div>

<h2>What's new?</h2>
    <p>Since our last paper we added quite some 
    new resources and features.
    Most importantly we largely extended the scope of the prior knowledge
    we cover, including annotations about protein functions and localizations,
    protein complexes and a comprehensive classification of protein roles
    in inter-cellular communication.
    Overall we build OmniPath using more than 115 databases.
    We also added homology translation to mouse and 
    rat, transcriptional regulation from <a target="_blank"
    href="https://saezlab.github.io/DoRothEA/">DoRothEA</a> and other
    databases, miRNA-mRNA interactions, a number of ligand-receptor
    databases, improved processing of enzyme-substrate interactions and 
    Gene Ontology, many other things. We even got a 
    <a href="https://apps.cytoscape.org/apps/omnipath" target="_blank">
    Cytoscape plug-in</a> thanks to
    <a href="https://www.linkedin.com/in/francesco-ceccarelli-41a660161/"
        target="_blank">Francesco Ceccarelli</a>. With
    <a href="https://saezlab.org/person/nicolas-palacio/"
        target="_blank">Nicolàs Palacio</a> and
    <a href="https://saezlab.org/person/olga-ivanova/"
        target="_blank">Olga Ivanova</a>
    we improve pypath code quality,
    documentation and tests.
    On the figure below you can see the new parts highlighted by violet. 
    Almost every couple of weeks we add new 
    ones, stay tuned and follow our <a href="https://twitter.com/omnipathdb"
        target="_blank">Twitter!</a>
    </p>
    <p>
        <a href="https://static.omnipathdb.org/pypath_overview.pdf">
        <img
        src="https://static.omnipathdb.org/pypath_overview.png"
        title="Overview of resources in pypath and OmniPath. 
            New additions since Nov 2016 highlighted in violet. 
            Expanded from Fig 1 of Turei et al. Nature Methods (2016)."
        style="width: 100%; max-width: 980px;"
        />
        </a>
    </p>

    <p style="font-size: small; max-width: 980px;"><em>
        Overview of resources in pypath and OmniPath.
        New additions since Nov 2016 highlighted in violet.
        Expanded from Fig 1 of Turei et al. Nature Methods (2016).
        For high resolution PDF
        <a href="https://static.omnipathdb.org/pypath_overview.pdf">
        click here</a>.
    </em></p>

<h2 id="paper">
    <a class="omnipath"
        href="https://www.nature.com/nmeth/journal/v13/n12/full/nmeth.4077.html"
        target="_blank">
        OmniPath publications
    </a>
</h2>

<p>
    D Turei, T Korcsmaros and J Saez-Rodriguez (2016) OmniPath: guidelines
    and gateway for literature-curated signaling pathway resources.
    <em>Nature Methods</em> 13 (12)
    <a target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/27898060">
        PMID: 27898060
    </a>
</p>

<a href="https://rdcu.be/m53B" target="_blank"
    title="Access the paper via readcube">
    <span class="paper-link">click here for the general paper</span>
</a>

<p>
    F Ceccarelli, D Turei, A Gabor, J Saez-Rodriguez (2020) Bringing Data
    From Curated Pathway Resources to Cytoscape With OmniPath.
    <em>Bioinformatics</em> 36 (1)
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/31886476" target="_blank">
        PMID: 31886476
    </a>
</p>


<a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btz968/5689817?guestAccessKey=c47f593a-0017-43c9-b980-962003c5c318"
    target="_blank"
    title="Access the paper from the Oxford University Press web page">
    <span class="paper-link">click here for the Cytoscape app paper</span>
</a>

<h2 id="data"><a class="omnipath"
    href="https://github.com/saezlab/pypath/blob/master/webservice.rst" 
    target="_blank">
    How to access the data?
    </a>
</h2>
<p>
    OmniPath is available via a web service, an
    <a href="https://github.com/saezlab/OmnipathR" target="_blank">
        R package
    </a> (<a href="https://saezlab.github.io/OmnipathR/" target="_blank">
        website
    </a>),
    a
    <a href="https://apps.cytoscape.org/apps/omnipath" target="_blank">
        Cytoscape plug-in
    </a>
    and a
    <a href="https://github.com/saezlab/pypath" target="_blank">
        Python module
    </a>.
    <a href="https://github.com/saezlab/pypath/blob/master/webservice.rst"
        target="_blank">
        For the web service see example queries here.
    </a>
</p>
<p>For old versions of the web service content see the
    <a href="https://archive.omnipathdb.org/" target="_blank">archive</a>.
</p>
<p>In the webservice you can access not only OmniPath but also TF-gene
    interactions from <a href="https://saezlab.github.io/DoRothEA/"
    target="_blank">DoRothEA</a>. See example queries in the readme at the
    <a href="https://github.com/saezlab/DoRothEA" target="_blank">DoRothEA</a>
    git repo.
    The web service provides 5 main query types: interactions,
    enzyme-substrate relationships, protein complexes, annotations and
    inter-cellular communication roles.
    <a href="https://github.com/saezlab/pypath/blob/master/webservice.rst"
        target="_blank">
        Here you can see examples for all of these query types.
    </a>
</p>

<h2 id="pypath">What you can do with pypath?</h2>

<p>
    <em>pypath</em> is a Python module which builds all of the five main
    database segments of OmniPath: the network, the enzyme-substrate
    relationships, the protein complexes, the annotations and the
    inter-cellular communication roles.
    If you use pypath instead of the web service you will be able (1) to
    customize the databases and (2) to access the data by a versatile and
    flexible API.
    In addition pypath provides many utility modules for example to convert
    between different identifiers, work with Gene Ontology, and many more.
</p>

<h2 id="code"><a class="omnipath" href="https://github.com/saezlab/pypath"
    target="_blank">pypath code</a>
</h2>

    <p>Pypath is available under GPLv3 license from.
        <a href="https://github.com/saezlab/pypath" target="_blank">
        this git repository.
        </a>
    </p>
    <p>
        <a href="https://github.com/saezlab/pypath/issues" target="_blank">
        Forks, pull requests, bug reports and feature requests are welcome!
        </a>
    </p>

<h3 id="install">Installation</h3>

    <p>
    <strong>New:</strong>
    Since the beginning of 2020 <em>igraph</em> and <em>cairo</em> are not
    mandatory dependencies of <em>pypath</em> which makes the installation
    process much easier on Mac.
    We will updates the guides below soon.
    </p>

<p><strong>Linux:</strong></p>
<div class="box codebox">
    pip install git+git://github.com/saezlab/pypath.git
</div>
<p><strong>OS X:</strong> Installation on Mac might be challenging, mainly 
    because of cairo. Follow the scripts below, go step-by-step and watch out 
    for errors. We tested these methods on several computers, but every 
    system is different. We appreciate if you report any error you 
    experience, and we will do our best to find a solution. Please contact 
    omnipath@googlegroups.com.
</p>
<ul>
    <li><a href="https://github.com/saezlab/pypath/blob/master/
        mac-install-brew.sh"
        target="_blank">With HomeBrew</a>
    </li>
    <li><a href="https://github.com/saezlab/pypath/blob/master/
        mac-install-conda.sh"
        target="_blank">With Anaconda</a>
    </li>
</ul>

<h2 id="docs">
    <a class="omnipath" href="https://pypath.omnipathdb.org/"
        target = "_blank">pypath documentation</a>
</h2>
 <p>For the reference documentation of the pypath module 
    <a href="https://pypath.omnipathdb.org/" target="_blank">click here.</a>
</p>

<h2 id="resources"><a class="omnipath" href="/info">
    Metainformation about signaling pathway resources</a>
</h2>
<p><a href="/info">Find here a comprehensive list of signaling pathway 
    resources.</a>
</p>

<h2 id="tutorial">Tutorials</h2>

<p>
<strong>New:</strong>
Since the beginning of 2020 <em>pypath</em> has a new API for the
network database.
This new API is in "beta" phase at the moment and not covered yet in the
<em>Pypath Guide</em> but we will include it soon.
</p>
<p>Pypath allows to access and integrate dozens of bioinformatics resources 
    in an easy way. Using the capabilities of Python you can handle them with 
    a lot of flexibility. Just pop up a Python shell, and type:
</p>

<div class="codebox box">
    from pypath import network<br>
    <br>
    op = network.Network.omnipath()<br>
</div>

<!--TUTORIALS-->

<h2 id="help">Help</h2>

<p>
    If you have a question or experiencing issues with the data or pypath
    please open an issue in
    <a href="https://github.com/saezlab/pypath/issues" target="_blank">
        the git repo of pypath
    </a>. We closed our
    <a href="https://groups.google.com/forum/#!forum/omnipath" target="_blank">
        support channel in Google Groups
    </a> to eliminate the redundancy with the issues on github. If you have
    a private message which is not of public interest you can send us an 
    email to the address below.
</p>

<h2 id="contact">Contact</h2>

<p>
    Contact us with private messages at omnipathdb@gmail.com.
    If you have a question or issue which might be of public interest
    please open an issue on 
    <a href="https://github.com/saezlab/pypath/issues" target="_blank">
        github
    </a>.
</p>



<h2>Miscellanous</h2>

<h3>OmniPath in meta-databases</h3>
<p>
<ul>
    <li><a href="https://omictools.com/omnipath-tool"
        target="_blank">OmicTools</a>
    </li>
    <li><a href="http://pathguide.org/fullrecord.php?organisms=all&
        availability=all&standards=all&order=alphabetic&DBID=582"
        target="_blank">PathGuide</a>
    </li>
</ul>

<h3>OmniPath press releases</h3>
<p>
<ul>
    <li><a href="https://www.ebi.ac.uk/about/news/press-releases/
        omnipath-sends-strong-signal" target="_blank">EMBL-EBI</a>
    </li>
    <li><a href="https://www.earlham.ac.uk/newsroom/
        omnipath-sends-strong-signal" target="_blank">Earlham Institute</a>
    </li>
    <li><a href="https://phys.org/news/
        2016-11-combining-pathway-resources-clearer-view.html"
        target="_blank">Phys.org</a>
    </li>
</ul>
</p>

<h3>Contributing groups</h3>
<p>
<ul>
    <li><a href="https://saezlab.org/" target="_blank">Saez Lab</a></li>
    <li><a href="https://www.earlham.ac.uk/korcsmaros-group" target="_blank">
        Korcsmaros Lab</a>
    </li>
</ul>
</p>

<h3><a href="https://scholar.google.com/scholar?oi=bibs&hl=en&cites=2131297002329856377" target="_blank">
    Papers citing OmniPath</a>
</h3>
