---

title: BioInformatics Tools kit


keywords: fastai
sidebar: home_sidebar

summary: "Summarized some useful tools in it."
description: "Summarized some useful tools in it."
nb_path: "index.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: index.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>BITK will help us manipulate data in different formats.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Install">Install<a class="anchor-link" href="#Install"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><code>pip install bitk</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="How-to-use">How to use<a class="anchor-link" href="#How-to-use"> </a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="dedim.py">dedim.py<a class="anchor-link" href="#dedim.py"> </a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<div class="highlight"><pre><span></span>usage: dedim.py <span class="o">[</span>-h<span class="o">]</span> <span class="o">[</span>--sep SEP<span class="o">]</span> <span class="o">[</span>--dedimensions-method DEDIMENSIONS_METHOD<span class="o">]</span>
                <span class="o">[</span>--cluster-method CLUSTER_METHOD<span class="o">]</span>
                <span class="o">[</span>--assess-method ASSESS_METHOD<span class="o">]</span> <span class="o">[</span>--dimensions DIMENSIONS<span class="o">]</span>
                <span class="o">[</span>--cluster-number CLUSTER_NUMBER<span class="o">]</span> <span class="o">[</span>-r<span class="o">]</span> <span class="o">[</span>--no-row-feature<span class="o">]</span>
                <span class="o">[</span>--annotation ANNOTATION<span class="o">]</span> <span class="o">[</span>--size SIZE<span class="o">]</span> <span class="o">[</span>--style STYLE<span class="o">]</span>
                <span class="o">[</span>-t TITLE<span class="o">]</span> <span class="o">[</span>-f FIG<span class="o">]</span> <span class="o">[</span>--version<span class="o">]</span>
                matrix prefix

positional arguments:
  matrix                matrix table, <span class="k">if</span> row represents feature, please note to add <span class="s1">&#39;--row-feature&#39;</span> option
  prefix                output prefix

optional arguments:
  -h, --help            show this <span class="nb">help</span> message and <span class="nb">exit</span>
  --sep SEP             separation
                        <span class="o">(</span>default:   <span class="o">)</span>
  --dedimensions-method DEDIMENSIONS_METHOD
                        de-dimensions method
                        <span class="o">(</span>default: PCA<span class="o">)</span>
  --cluster-method CLUSTER_METHOD
                        cluster method
                        <span class="o">(</span>default: MiniBatchKMeans<span class="o">)</span>
  --assess-method ASSESS_METHOD
                        assess methods <span class="k">for</span> best cluster number
                        <span class="o">(</span>default: silhouette_score<span class="o">)</span>
  --dimensions DIMENSIONS
                        reduce to n dimensions
                        <span class="o">(</span>default: <span class="m">3</span><span class="o">)</span>
  --cluster-number CLUSTER_NUMBER
                        cluster number, <span class="k">if</span> not specific it, it will be the best cluster number infered
                        <span class="o">(</span>default: None<span class="o">)</span>
  -r, --row-feature     row in the matrix represents feature
                        <span class="o">(</span>default: True<span class="o">)</span>
  --no-row-feature
  --annotation ANNOTATION
                        annotation file, sep should be <span class="s1">&#39;,&#39;</span>
                        <span class="o">(</span>default: None<span class="o">)</span>
  --size SIZE           size column in annotation file
                        <span class="o">(</span>default: None<span class="o">)</span>
  --style STYLE         style column in annotation file
                        <span class="o">(</span>default: None<span class="o">)</span>
  -t TITLE, --title TITLE
                        figure title
                        <span class="o">(</span>default: None<span class="o">)</span>
  -f FIG, --fig FIG     png/pdf
                        <span class="o">(</span>default: png<span class="o">)</span>
  --version             show program<span class="err">&#39;</span>s version number and <span class="nb">exit</span>
</pre></div>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="example">example<a class="anchor-link" href="#example"> </a></h4><div class="highlight"><pre><span></span>dedim.py tests/test.csv tests/test_result <span class="se">\</span>
        --sep , --dimensions <span class="m">2</span> <span class="se">\</span>
        --no-row-feature --annotation tests/test_anno.csv <span class="se">\</span>
        --style targets --title <span class="s1">&#39;test PCA&#39;</span> <span class="se">\</span>
        --dedimensions-method TSNE --fig png
</pre></div>
<p><img src="/simplebitk/tests/test_result_TSNE.png" alt="pca"></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="fc_rename.py">fc_rename.py<a class="anchor-link" href="#fc_rename.py"> </a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

<pre><code>usage: fc_rename.py [-h] [-s SAMPLE_TITLE] [-b BAM_TITLE] [-c COUNT_TITLE]
                    [--version]
                    featurecounts clinical prefix

featurecounts will use the bamfile name as the sample name. This script rename the featurecounts by the given table.

positional arguments:
  featurecounts         featurecounts table file
  clinical              clinical table file with header
  prefix                output prefix, there will be two outputs, one is count output (prefix_count.txt), and the other one is the rename featurecounts (prefix_fc.txt)

optional arguments:
  -h, --help            show this help message and exit
  -s SAMPLE_TITLE, --sample-title SAMPLE_TITLE
                        the column name of the sample name
                        (default: sample)
  -b BAM_TITLE, --bam-title BAM_TITLE
                        the column name of the bamfile name
                        (default: bam)
  -c COUNT_TITLE, --count-title COUNT_TITLE
                        the column name used as identity in count data
                        (default: Geneid)
  --version             show program's version number and exit</code></pre>

</div>
</div>
</div>
</div>
 

