---

title: preprocess


keywords: fastai
sidebar: home_sidebar



nb_path: "06_preprocess.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 06_preprocess.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fc_count2tpm" class="doc_header"><code>fc_count2tpm</code><a href="https://github.com/btrspg/simplebitk/tree/master/simplebitk/preprocess.py#L11" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fc_count2tpm</code>(<strong><code>df</code></strong>, <strong><code>samples</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fc_count2fpkm" class="doc_header"><code>fc_count2fpkm</code><a href="https://github.com/btrspg/simplebitk/tree/master/simplebitk/preprocess.py#L17" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fc_count2fpkm</code>(<strong><code>df</code></strong>, <strong><code>samples</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="featurecount_rename" class="doc_header"><code>featurecount_rename</code><a href="https://github.com/btrspg/simplebitk/tree/master/simplebitk/preprocess.py#L28" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>featurecount_rename</code>(<strong><code>featurecounts</code></strong>, <strong><code>clinical</code></strong>, <strong><code>prefix</code></strong>, <strong><code>sample_title</code></strong>=<em><code>'sample'</code></em>, <strong><code>bam_title</code></strong>=<em><code>'bam'</code></em>, <strong><code>count_title</code></strong>=<em><code>'Geneid'</code></em>)</p>
</blockquote>
<p>featurecounts will use the bamfile name as the sample name. This script rename the featurecounts by the given table.</p>
<p>:param str featurecounts: featurecounts table file
:param str clinical: clinical table file with header
:param str prefix: output prefix, there will be four outputs, one is count output (prefix_count.txt), and the other one is the rename featurecounts (prefix_featurecounts.txt) and FPKM and TPM {prefix}_featurecounts_FPKM.txt {prefix}_featurecounts_TPM.txt
:param str sample_title: the column name of the sample name
:param str bam_title: the column name of the bamfile name
:param str count_title: the column name used as identity in count data</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">featurecounts</span><span class="o">=</span><span class="s1">&#39;tests/fc.txt&#39;</span>
<span class="n">clinical</span><span class="o">=</span><span class="s1">&#39;tests/clinical.txt&#39;</span>
<span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;tests/rename&#39;</span>
<span class="n">sample_title</span><span class="o">=</span><span class="s1">&#39;sample&#39;</span>
<span class="n">bam_title</span><span class="o">=</span><span class="s1">&#39;starbam&#39;</span>
<span class="n">count_title</span><span class="o">=</span><span class="s1">&#39;gene_name&#39;</span>
<span class="n">featurecount_rename</span><span class="p">(</span><span class="n">featurecounts</span><span class="p">,</span><span class="n">clinical</span><span class="p">,</span><span class="n">prefix</span><span class="p">,</span><span class="n">sample_title</span><span class="o">=</span><span class="n">sample_title</span><span class="p">,</span><span class="n">bam_title</span><span class="o">=</span><span class="n">bam_title</span><span class="p">,</span><span class="n">count_title</span><span class="o">=</span><span class="n">count_title</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="o">!</span>head tests/fc.txt tests/clinical.txt tests/rename*
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>==&gt; tests/fc.txt &lt;==
#Program:featureCounts	v2.0.1;	Command:&#34;featureCounts&#34;	&#34;-T&#34;	&#34;12&#34;	&#34;-p&#34;	&#34;-t&#34;	&#34;exon&#34;	&#34;-g&#34;	&#34;gene_id&#34;	&#34;-M&#34;	&#34;--ignoreDup&#34;	&#34;--extraAttributes&#34;	&#34;gene_name,gene_biotype&#34;	&#34;-a&#34;	&#34;/aegis/database/mus_musculus/99/transcriptome.gtf&#34;	&#34;-o&#34;	&#34;pair-end/results/RNA/merged/star_featurecounts.txt&#34;	&#34;--tmpDir&#34;	&#34;pair-end/results/RNA/tmp&#34;	&#34;pair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam&#34;	&#34;pair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam&#34;	&#34;pair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam&#34;	&#34;pair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam&#34;	&#34;pair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam&#34;	&#34;pair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam&#34;	
Geneid	Chr	Start	End	Strand	Length	gene_name	gene_biotype	pair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam	pair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam	pair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam	pair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam	pair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam	pair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam
ENSMUSG00000102693	1	3073253	3074322	+	1070	4933401J01Rik	TEC	0	0	0	0	0	0
ENSMUSG00000064842	1	3102016	3102125	+	110	Gm26206	snRNA	0	0	0	0	0	0
ENSMUSG00000051951	1;1;1;1;1;1;1	3205901;3206523;3213439;3213609;3214482;3421702;3670552	3207317;3207317;3215632;3216344;3216968;3421901;3671498	-;-;-;-;-;-;-	6094	Xkr4	protein_coding	6	9	6	13	12	11
ENSMUSG00000102851	1	3252757	3253236	+	480	Gm18956	processed_pseudogene	0	0	0	0	0	0
ENSMUSG00000103377	1	3365731	3368549	-	2819	Gm37180	TEC	0	0	0	0	0	0
ENSMUSG00000104017	1	3375556	3377788	-	2233	Gm37363	TEC	0	0	0	0	0	0
==&gt; tests/clinical.txt &lt;==
sample,condition,batch,starbam,hisat2bam
SRR12678525,Control,b1,pair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678525/align_data/SRR12678525_hisat2.bam
SRR12678526,Control,b1,pair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678526/align_data/SRR12678526_hisat2.bam
SRR12678527,Control,b1,pair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678527/align_data/SRR12678527_hisat2.bam
SRR12678528,Case,b1,pair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678528/align_data/SRR12678528_hisat2.bam
SRR12678529,Case,b1,pair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678529/align_data/SRR12678529_hisat2.bam
SRR12678530,Case,b1,pair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678530/align_data/SRR12678530_hisat2.bam
==&gt; tests/rename_counts.txt &lt;==
gene_name	SRR12678525	SRR12678526	SRR12678527	SRR12678528	SRR12678529	SRR12678530
4933401J01Rik	0	0	0	0	0	0
Gm26206	0	0	0	0	0	0
Xkr4	6	9	6	13	12	11
Gm18956	0	0	0	0	0	0
Gm37180	0	0	0	0	0	0
Gm37363	0	0	0	0	0	0

==&gt; tests/rename_fc.txt &lt;==
Geneid	Chr	Start	End	Strand	Length	gene_name	gene_biotype	SRR12678525	SRR12678526	SRR12678527	SRR12678528	SRR12678529	SRR12678530
ENSMUSG00000102693	1	3073253	3074322	+	1070	4933401J01Rik	TEC	0	0	0	0	0	0
ENSMUSG00000064842	1	3102016	3102125	+	110	Gm26206	snRNA	0	0	0	0	0	0
ENSMUSG00000051951	1;1;1;1;1;1;1	3205901;3206523;3213439;3213609;3214482;3421702;3670552	3207317;3207317;3215632;3216344;3216968;3421901;3671498	-;-;-;-;-;-;-	6094	Xkr4	protein_coding	6	9	6	13	12	11
ENSMUSG00000102851	1	3252757	3253236	+	480	Gm18956	processed_pseudogene	0	0	0	0	0	0
ENSMUSG00000103377	1	3365731	3368549	-	2819	Gm37180	TEC	0	0	0	0	0	0
ENSMUSG00000104017	1	3375556	3377788	-	2233	Gm37363	TEC	0	0	0	0	0	0

==&gt; tests/rename_featurecounts.txt &lt;==
Geneid	Chr	Start	End	Strand	Length	gene_name	gene_biotype	SRR12678525	SRR12678526	SRR12678527	SRR12678528	SRR12678529	SRR12678530
ENSMUSG00000102693	1	3073253	3074322	+	1070	4933401J01Rik	TEC	0	0	0	0	0	0
ENSMUSG00000064842	1	3102016	3102125	+	110	Gm26206	snRNA	0	0	0	0	0	0
ENSMUSG00000051951	1;1;1;1;1;1;1	3205901;3206523;3213439;3213609;3214482;3421702;3670552	3207317;3207317;3215632;3216344;3216968;3421901;3671498	-;-;-;-;-;-;-	6094	Xkr4	protein_coding	6	9	6	13	12	11
ENSMUSG00000102851	1	3252757	3253236	+	480	Gm18956	processed_pseudogene	0	0	0	0	0	0
ENSMUSG00000103377	1	3365731	3368549	-	2819	Gm37180	TEC	0	0	0	0	0	0
ENSMUSG00000104017	1	3375556	3377788	-	2233	Gm37363	TEC	0	0	0	0	0	0

==&gt; tests/rename_featurecounts_FPKM.txt &lt;==
Geneid	Chr	Start	End	Strand	Length	gene_name	gene_biotype	SRR12678525	SRR12678526	SRR12678527	SRR12678528	SRR12678529	SRR12678530
ENSMUSG00000102693	1	3073253	3074322	+	1070	4933401J01Rik	TEC	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000064842	1	3102016	3102125	+	110	Gm26206	snRNA	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000051951	1;1;1;1;1;1;1	3205901;3206523;3213439;3213609;3214482;3421702;3670552	3207317;3207317;3215632;3216344;3216968;3421901;3671498	-;-;-;-;-;-;-	6094	Xkr4	protein_coding	164.0958319658681	164.09583196586806	164.0958319658681	164.0958319658681	164.0958319658681	164.0958319658681
ENSMUSG00000102851	1	3252757	3253236	+	480	Gm18956	processed_pseudogene	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000103377	1	3365731	3368549	-	2819	Gm37180	TEC	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000104017	1	3375556	3377788	-	2233	Gm37363	TEC	0.0	0.0	0.0	0.0	0.0	0.0

==&gt; tests/rename_featurecounts_TPM.txt &lt;==
Geneid	Chr	Start	End	Strand	Length	gene_name	gene_biotype	SRR12678525	SRR12678526	SRR12678527	SRR12678528	SRR12678529	SRR12678530
ENSMUSG00000102693	1	3073253	3074322	+	1070	4933401J01Rik	TEC	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000064842	1	3102016	3102125	+	110	Gm26206	snRNA	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000051951	1;1;1;1;1;1;1	3205901;3206523;3213439;3213609;3214482;3421702;3670552	3207317;3207317;3215632;3216344;3216968;3421901;3671498	-;-;-;-;-;-;-	6094	Xkr4	protein_coding	1000000.0	1000000.0	1000000.0	1000000.0	1000000.0	1000000.0
ENSMUSG00000102851	1	3252757	3253236	+	480	Gm18956	processed_pseudogene	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000103377	1	3365731	3368549	-	2819	Gm37180	TEC	0.0	0.0	0.0	0.0	0.0	0.0
ENSMUSG00000104017	1	3375556	3377788	-	2233	Gm37363	TEC	0.0	0.0	0.0	0.0	0.0	0.0
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">prefix</span><span class="o">+</span><span class="s1">&#39;_featurecounts.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;SRR12678525&#39;</span><span class="p">,</span> <span class="s1">&#39;SRR12678526&#39;</span><span class="p">,</span> <span class="s1">&#39;SRR12678527&#39;</span><span class="p">,</span> <span class="s1">&#39;SRR12678528&#39;</span><span class="p">,</span> <span class="s1">&#39;SRR12678529&#39;</span><span class="p">,</span> <span class="s1">&#39;SRR12678530&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">fc_count2tpm</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">samples</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Geneid</th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Strand</th>
      <th>Length</th>
      <th>gene_name</th>
      <th>gene_biotype</th>
      <th>SRR12678525</th>
      <th>SRR12678526</th>
      <th>SRR12678527</th>
      <th>SRR12678528</th>
      <th>SRR12678529</th>
      <th>SRR12678530</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENSMUSG00000102693</td>
      <td>1</td>
      <td>3073253</td>
      <td>3074322</td>
      <td>+</td>
      <td>1070</td>
      <td>4933401J01Rik</td>
      <td>TEC</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSMUSG00000064842</td>
      <td>1</td>
      <td>3102016</td>
      <td>3102125</td>
      <td>+</td>
      <td>110</td>
      <td>Gm26206</td>
      <td>snRNA</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSMUSG00000051951</td>
      <td>1;1;1;1;1;1;1</td>
      <td>3205901;3206523;3213439;3213609;3214482;342170...</td>
      <td>3207317;3207317;3215632;3216344;3216968;342190...</td>
      <td>-;-;-;-;-;-;-</td>
      <td>6094</td>
      <td>Xkr4</td>
      <td>protein_coding</td>
      <td>1000000.0</td>
      <td>1000000.0</td>
      <td>1000000.0</td>
      <td>1000000.0</td>
      <td>1000000.0</td>
      <td>1000000.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSMUSG00000102851</td>
      <td>1</td>
      <td>3252757</td>
      <td>3253236</td>
      <td>+</td>
      <td>480</td>
      <td>Gm18956</td>
      <td>processed_pseudogene</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSMUSG00000103377</td>
      <td>1</td>
      <td>3365731</td>
      <td>3368549</td>
      <td>-</td>
      <td>2819</td>
      <td>Gm37180</td>
      <td>TEC</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>ENSMUSG00000104017</td>
      <td>1</td>
      <td>3375556</td>
      <td>3377788</td>
      <td>-</td>
      <td>2233</td>
      <td>Gm37363</td>
      <td>TEC</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">fc_count2fpkm</span><span class="p">(</span><span class="n">df</span><span class="p">,</span><span class="n">samples</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Geneid</th>
      <th>Chr</th>
      <th>Start</th>
      <th>End</th>
      <th>Strand</th>
      <th>Length</th>
      <th>gene_name</th>
      <th>gene_biotype</th>
      <th>SRR12678525</th>
      <th>SRR12678526</th>
      <th>SRR12678527</th>
      <th>SRR12678528</th>
      <th>SRR12678529</th>
      <th>SRR12678530</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENSMUSG00000102693</td>
      <td>1</td>
      <td>3073253</td>
      <td>3074322</td>
      <td>+</td>
      <td>1070</td>
      <td>4933401J01Rik</td>
      <td>TEC</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSMUSG00000064842</td>
      <td>1</td>
      <td>3102016</td>
      <td>3102125</td>
      <td>+</td>
      <td>110</td>
      <td>Gm26206</td>
      <td>snRNA</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSMUSG00000051951</td>
      <td>1;1;1;1;1;1;1</td>
      <td>3205901;3206523;3213439;3213609;3214482;342170...</td>
      <td>3207317;3207317;3215632;3216344;3216968;342190...</td>
      <td>-;-;-;-;-;-;-</td>
      <td>6094</td>
      <td>Xkr4</td>
      <td>protein_coding</td>
      <td>164.095832</td>
      <td>164.095832</td>
      <td>164.095832</td>
      <td>164.095832</td>
      <td>164.095832</td>
      <td>164.095832</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSMUSG00000102851</td>
      <td>1</td>
      <td>3252757</td>
      <td>3253236</td>
      <td>+</td>
      <td>480</td>
      <td>Gm18956</td>
      <td>processed_pseudogene</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSMUSG00000103377</td>
      <td>1</td>
      <td>3365731</td>
      <td>3368549</td>
      <td>-</td>
      <td>2819</td>
      <td>Gm37180</td>
      <td>TEC</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>5</th>
      <td>ENSMUSG00000104017</td>
      <td>1</td>
      <td>3375556</td>
      <td>3377788</td>
      <td>-</td>
      <td>2233</td>
      <td>Gm37363</td>
      <td>TEC</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

