{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp preprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def fc_count2tpm(df,samples):\n",
    "    results = df.copy()\n",
    "    rate=results[samples].divide(results['Length'],axis=0)\n",
    "    results[samples]=rate/rate.sum(axis=0)*1e6\n",
    "    return results\n",
    "\n",
    "def fc_count2fpkm(df,samples):\n",
    "    results = df.copy()\n",
    "    total_reads = results[samples].sum(axis=0)\n",
    "    rate = results[samples].divide(results['Length'],axis=0)\n",
    "    results[samples] = rate.divide(total_reads,axis=1) * 1e6\n",
    "    return results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export\n",
    "\n",
    "def featurecount_rename(featurecounts,clinical,prefix,*,sample_title='sample',bam_title='bam',count_title='Geneid'):\n",
    "    '''\n",
    "    featurecounts will use the bamfile name as the sample name. This script rename the featurecounts by the given table.\n",
    "    \n",
    "    :param str featurecounts: featurecounts table file\n",
    "    :param str clinical: clinical table file with header\n",
    "    :param str prefix: output prefix, there will be four outputs, one is count output (prefix_count.txt), and the other one is the rename featurecounts (prefix_featurecounts.txt) and FPKM and TPM {prefix}_featurecounts_FPKM.txt {prefix}_featurecounts_TPM.txt\n",
    "    :param str sample_title: the column name of the sample name\n",
    "    :param str bam_title: the column name of the bamfile name\n",
    "    :param str count_title: the column name used as identity in count data\n",
    "    '''\n",
    "    \n",
    "    fc = pd.read_csv(featurecounts,sep='\\t',comment='#')\n",
    "    cli = pd.read_csv(clinical,comment='#')\n",
    "    fc.rename(columns=cli.set_index(bam_title)[sample_title], \n",
    "                 inplace=True)\n",
    "    samples=list(cli[sample_title])\n",
    "    fc.to_csv(f'{prefix}_featurecounts.txt',sep='\\t',index=False)\n",
    "    fc[[count_title]+samples].to_csv(f'{prefix}_counts.txt',sep='\\t',index=False)\n",
    "    fpkm=fc_count2fpkm(fc,samples)\n",
    "    tpm=fc_count2tpm(fc,samples)\n",
    "    fpkm.to_csv(f'{prefix}_featurecounts_FPKM.txt',sep='\\t',index=False)\n",
    "    tpm.to_csv(f'{prefix}_featurecounts_TPM.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "featurecounts='tests/fc.txt'\n",
    "clinical='tests/clinical.txt'\n",
    "prefix='tests/rename'\n",
    "sample_title='sample'\n",
    "bam_title='starbam'\n",
    "count_title='gene_name'\n",
    "featurecount_rename(featurecounts,clinical,prefix,sample_title=sample_title,bam_title=bam_title,count_title=count_title)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> tests/fc.txt <==\n",
      "#Program:featureCounts\tv2.0.1;\tCommand:\"featureCounts\"\t\"-T\"\t\"12\"\t\"-p\"\t\"-t\"\t\"exon\"\t\"-g\"\t\"gene_id\"\t\"-M\"\t\"--ignoreDup\"\t\"--extraAttributes\"\t\"gene_name,gene_biotype\"\t\"-a\"\t\"/aegis/database/mus_musculus/99/transcriptome.gtf\"\t\"-o\"\t\"pair-end/results/RNA/merged/star_featurecounts.txt\"\t\"--tmpDir\"\t\"pair-end/results/RNA/tmp\"\t\"pair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam\"\t\"pair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam\"\t\"pair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam\"\t\"pair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam\"\t\"pair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam\"\t\"pair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam\"\t\n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\tgene_name\tgene_biotype\tpair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam\tpair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam\tpair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam\tpair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam\tpair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam\tpair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam\n",
      "ENSMUSG00000102693\t1\t3073253\t3074322\t+\t1070\t4933401J01Rik\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000064842\t1\t3102016\t3102125\t+\t110\tGm26206\tsnRNA\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000051951\t1;1;1;1;1;1;1\t3205901;3206523;3213439;3213609;3214482;3421702;3670552\t3207317;3207317;3215632;3216344;3216968;3421901;3671498\t-;-;-;-;-;-;-\t6094\tXkr4\tprotein_coding\t6\t9\t6\t13\t12\t11\n",
      "ENSMUSG00000102851\t1\t3252757\t3253236\t+\t480\tGm18956\tprocessed_pseudogene\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000103377\t1\t3365731\t3368549\t-\t2819\tGm37180\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000104017\t1\t3375556\t3377788\t-\t2233\tGm37363\tTEC\t0\t0\t0\t0\t0\t0\n",
      "==> tests/clinical.txt <==\n",
      "sample,condition,batch,starbam,hisat2bam\n",
      "SRR12678525,Control,b1,pair-end/results/RNA/SRR12678525/align_data/SRR12678525_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678525/align_data/SRR12678525_hisat2.bam\n",
      "SRR12678526,Control,b1,pair-end/results/RNA/SRR12678526/align_data/SRR12678526_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678526/align_data/SRR12678526_hisat2.bam\n",
      "SRR12678527,Control,b1,pair-end/results/RNA/SRR12678527/align_data/SRR12678527_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678527/align_data/SRR12678527_hisat2.bam\n",
      "SRR12678528,Case,b1,pair-end/results/RNA/SRR12678528/align_data/SRR12678528_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678528/align_data/SRR12678528_hisat2.bam\n",
      "SRR12678529,Case,b1,pair-end/results/RNA/SRR12678529/align_data/SRR12678529_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678529/align_data/SRR12678529_hisat2.bam\n",
      "SRR12678530,Case,b1,pair-end/results/RNA/SRR12678530/align_data/SRR12678530_STAR_Aligned.sortedByCoord.out.bam,pair-end/results/RNA/SRR12678530/align_data/SRR12678530_hisat2.bam\n",
      "==> tests/rename_counts.txt <==\n",
      "gene_name\tSRR12678525\tSRR12678526\tSRR12678527\tSRR12678528\tSRR12678529\tSRR12678530\n",
      "4933401J01Rik\t0\t0\t0\t0\t0\t0\n",
      "Gm26206\t0\t0\t0\t0\t0\t0\n",
      "Xkr4\t6\t9\t6\t13\t12\t11\n",
      "Gm18956\t0\t0\t0\t0\t0\t0\n",
      "Gm37180\t0\t0\t0\t0\t0\t0\n",
      "Gm37363\t0\t0\t0\t0\t0\t0\n",
      "\n",
      "==> tests/rename_fc.txt <==\n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\tgene_name\tgene_biotype\tSRR12678525\tSRR12678526\tSRR12678527\tSRR12678528\tSRR12678529\tSRR12678530\n",
      "ENSMUSG00000102693\t1\t3073253\t3074322\t+\t1070\t4933401J01Rik\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000064842\t1\t3102016\t3102125\t+\t110\tGm26206\tsnRNA\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000051951\t1;1;1;1;1;1;1\t3205901;3206523;3213439;3213609;3214482;3421702;3670552\t3207317;3207317;3215632;3216344;3216968;3421901;3671498\t-;-;-;-;-;-;-\t6094\tXkr4\tprotein_coding\t6\t9\t6\t13\t12\t11\n",
      "ENSMUSG00000102851\t1\t3252757\t3253236\t+\t480\tGm18956\tprocessed_pseudogene\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000103377\t1\t3365731\t3368549\t-\t2819\tGm37180\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000104017\t1\t3375556\t3377788\t-\t2233\tGm37363\tTEC\t0\t0\t0\t0\t0\t0\n",
      "\n",
      "==> tests/rename_featurecounts.txt <==\n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\tgene_name\tgene_biotype\tSRR12678525\tSRR12678526\tSRR12678527\tSRR12678528\tSRR12678529\tSRR12678530\n",
      "ENSMUSG00000102693\t1\t3073253\t3074322\t+\t1070\t4933401J01Rik\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000064842\t1\t3102016\t3102125\t+\t110\tGm26206\tsnRNA\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000051951\t1;1;1;1;1;1;1\t3205901;3206523;3213439;3213609;3214482;3421702;3670552\t3207317;3207317;3215632;3216344;3216968;3421901;3671498\t-;-;-;-;-;-;-\t6094\tXkr4\tprotein_coding\t6\t9\t6\t13\t12\t11\n",
      "ENSMUSG00000102851\t1\t3252757\t3253236\t+\t480\tGm18956\tprocessed_pseudogene\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000103377\t1\t3365731\t3368549\t-\t2819\tGm37180\tTEC\t0\t0\t0\t0\t0\t0\n",
      "ENSMUSG00000104017\t1\t3375556\t3377788\t-\t2233\tGm37363\tTEC\t0\t0\t0\t0\t0\t0\n",
      "\n",
      "==> tests/rename_featurecounts_FPKM.txt <==\n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\tgene_name\tgene_biotype\tSRR12678525\tSRR12678526\tSRR12678527\tSRR12678528\tSRR12678529\tSRR12678530\n",
      "ENSMUSG00000102693\t1\t3073253\t3074322\t+\t1070\t4933401J01Rik\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000064842\t1\t3102016\t3102125\t+\t110\tGm26206\tsnRNA\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000051951\t1;1;1;1;1;1;1\t3205901;3206523;3213439;3213609;3214482;3421702;3670552\t3207317;3207317;3215632;3216344;3216968;3421901;3671498\t-;-;-;-;-;-;-\t6094\tXkr4\tprotein_coding\t164.0958319658681\t164.09583196586806\t164.0958319658681\t164.0958319658681\t164.0958319658681\t164.0958319658681\n",
      "ENSMUSG00000102851\t1\t3252757\t3253236\t+\t480\tGm18956\tprocessed_pseudogene\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000103377\t1\t3365731\t3368549\t-\t2819\tGm37180\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000104017\t1\t3375556\t3377788\t-\t2233\tGm37363\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "\n",
      "==> tests/rename_featurecounts_TPM.txt <==\n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\tgene_name\tgene_biotype\tSRR12678525\tSRR12678526\tSRR12678527\tSRR12678528\tSRR12678529\tSRR12678530\n",
      "ENSMUSG00000102693\t1\t3073253\t3074322\t+\t1070\t4933401J01Rik\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000064842\t1\t3102016\t3102125\t+\t110\tGm26206\tsnRNA\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000051951\t1;1;1;1;1;1;1\t3205901;3206523;3213439;3213609;3214482;3421702;3670552\t3207317;3207317;3215632;3216344;3216968;3421901;3671498\t-;-;-;-;-;-;-\t6094\tXkr4\tprotein_coding\t1000000.0\t1000000.0\t1000000.0\t1000000.0\t1000000.0\t1000000.0\n",
      "ENSMUSG00000102851\t1\t3252757\t3253236\t+\t480\tGm18956\tprocessed_pseudogene\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000103377\t1\t3365731\t3368549\t-\t2819\tGm37180\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n",
      "ENSMUSG00000104017\t1\t3375556\t3377788\t-\t2233\tGm37363\tTEC\t0.0\t0.0\t0.0\t0.0\t0.0\t0.0\n"
     ]
    }
   ],
   "source": [
    "!head tests/fc.txt tests/clinical.txt tests/rename*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.read_csv(prefix+'_featurecounts.txt',sep='\\t')\n",
    "samples=['SRR12678525', 'SRR12678526', 'SRR12678527', 'SRR12678528', 'SRR12678529', 'SRR12678530']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>SRR12678525</th>\n",
       "      <th>SRR12678526</th>\n",
       "      <th>SRR12678527</th>\n",
       "      <th>SRR12678528</th>\n",
       "      <th>SRR12678529</th>\n",
       "      <th>SRR12678530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>1070</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>1</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>110</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>1;1;1;1;1;1;1</td>\n",
       "      <td>3205901;3206523;3213439;3213609;3214482;342170...</td>\n",
       "      <td>3207317;3207317;3215632;3216344;3216968;342190...</td>\n",
       "      <td>-;-;-;-;-;-;-</td>\n",
       "      <td>6094</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000102851</td>\n",
       "      <td>1</td>\n",
       "      <td>3252757</td>\n",
       "      <td>3253236</td>\n",
       "      <td>+</td>\n",
       "      <td>480</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000103377</td>\n",
       "      <td>1</td>\n",
       "      <td>3365731</td>\n",
       "      <td>3368549</td>\n",
       "      <td>-</td>\n",
       "      <td>2819</td>\n",
       "      <td>Gm37180</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000104017</td>\n",
       "      <td>1</td>\n",
       "      <td>3375556</td>\n",
       "      <td>3377788</td>\n",
       "      <td>-</td>\n",
       "      <td>2233</td>\n",
       "      <td>Gm37363</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Geneid            Chr  \\\n",
       "0  ENSMUSG00000102693              1   \n",
       "1  ENSMUSG00000064842              1   \n",
       "2  ENSMUSG00000051951  1;1;1;1;1;1;1   \n",
       "3  ENSMUSG00000102851              1   \n",
       "4  ENSMUSG00000103377              1   \n",
       "5  ENSMUSG00000104017              1   \n",
       "\n",
       "                                               Start  \\\n",
       "0                                            3073253   \n",
       "1                                            3102016   \n",
       "2  3205901;3206523;3213439;3213609;3214482;342170...   \n",
       "3                                            3252757   \n",
       "4                                            3365731   \n",
       "5                                            3375556   \n",
       "\n",
       "                                                 End         Strand  Length  \\\n",
       "0                                            3074322              +    1070   \n",
       "1                                            3102125              +     110   \n",
       "2  3207317;3207317;3215632;3216344;3216968;342190...  -;-;-;-;-;-;-    6094   \n",
       "3                                            3253236              +     480   \n",
       "4                                            3368549              -    2819   \n",
       "5                                            3377788              -    2233   \n",
       "\n",
       "       gene_name          gene_biotype  SRR12678525  SRR12678526  SRR12678527  \\\n",
       "0  4933401J01Rik                   TEC          0.0          0.0          0.0   \n",
       "1        Gm26206                 snRNA          0.0          0.0          0.0   \n",
       "2           Xkr4        protein_coding    1000000.0    1000000.0    1000000.0   \n",
       "3        Gm18956  processed_pseudogene          0.0          0.0          0.0   \n",
       "4        Gm37180                   TEC          0.0          0.0          0.0   \n",
       "5        Gm37363                   TEC          0.0          0.0          0.0   \n",
       "\n",
       "   SRR12678528  SRR12678529  SRR12678530  \n",
       "0          0.0          0.0          0.0  \n",
       "1          0.0          0.0          0.0  \n",
       "2    1000000.0    1000000.0    1000000.0  \n",
       "3          0.0          0.0          0.0  \n",
       "4          0.0          0.0          0.0  \n",
       "5          0.0          0.0          0.0  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc_count2tpm(df,samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>SRR12678525</th>\n",
       "      <th>SRR12678526</th>\n",
       "      <th>SRR12678527</th>\n",
       "      <th>SRR12678528</th>\n",
       "      <th>SRR12678529</th>\n",
       "      <th>SRR12678530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>1</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>+</td>\n",
       "      <td>1070</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>1</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>+</td>\n",
       "      <td>110</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>1;1;1;1;1;1;1</td>\n",
       "      <td>3205901;3206523;3213439;3213609;3214482;342170...</td>\n",
       "      <td>3207317;3207317;3215632;3216344;3216968;342190...</td>\n",
       "      <td>-;-;-;-;-;-;-</td>\n",
       "      <td>6094</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>164.095832</td>\n",
       "      <td>164.095832</td>\n",
       "      <td>164.095832</td>\n",
       "      <td>164.095832</td>\n",
       "      <td>164.095832</td>\n",
       "      <td>164.095832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000102851</td>\n",
       "      <td>1</td>\n",
       "      <td>3252757</td>\n",
       "      <td>3253236</td>\n",
       "      <td>+</td>\n",
       "      <td>480</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000103377</td>\n",
       "      <td>1</td>\n",
       "      <td>3365731</td>\n",
       "      <td>3368549</td>\n",
       "      <td>-</td>\n",
       "      <td>2819</td>\n",
       "      <td>Gm37180</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000104017</td>\n",
       "      <td>1</td>\n",
       "      <td>3375556</td>\n",
       "      <td>3377788</td>\n",
       "      <td>-</td>\n",
       "      <td>2233</td>\n",
       "      <td>Gm37363</td>\n",
       "      <td>TEC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Geneid            Chr  \\\n",
       "0  ENSMUSG00000102693              1   \n",
       "1  ENSMUSG00000064842              1   \n",
       "2  ENSMUSG00000051951  1;1;1;1;1;1;1   \n",
       "3  ENSMUSG00000102851              1   \n",
       "4  ENSMUSG00000103377              1   \n",
       "5  ENSMUSG00000104017              1   \n",
       "\n",
       "                                               Start  \\\n",
       "0                                            3073253   \n",
       "1                                            3102016   \n",
       "2  3205901;3206523;3213439;3213609;3214482;342170...   \n",
       "3                                            3252757   \n",
       "4                                            3365731   \n",
       "5                                            3375556   \n",
       "\n",
       "                                                 End         Strand  Length  \\\n",
       "0                                            3074322              +    1070   \n",
       "1                                            3102125              +     110   \n",
       "2  3207317;3207317;3215632;3216344;3216968;342190...  -;-;-;-;-;-;-    6094   \n",
       "3                                            3253236              +     480   \n",
       "4                                            3368549              -    2819   \n",
       "5                                            3377788              -    2233   \n",
       "\n",
       "       gene_name          gene_biotype  SRR12678525  SRR12678526  SRR12678527  \\\n",
       "0  4933401J01Rik                   TEC     0.000000     0.000000     0.000000   \n",
       "1        Gm26206                 snRNA     0.000000     0.000000     0.000000   \n",
       "2           Xkr4        protein_coding   164.095832   164.095832   164.095832   \n",
       "3        Gm18956  processed_pseudogene     0.000000     0.000000     0.000000   \n",
       "4        Gm37180                   TEC     0.000000     0.000000     0.000000   \n",
       "5        Gm37363                   TEC     0.000000     0.000000     0.000000   \n",
       "\n",
       "   SRR12678528  SRR12678529  SRR12678530  \n",
       "0     0.000000     0.000000     0.000000  \n",
       "1     0.000000     0.000000     0.000000  \n",
       "2   164.095832   164.095832   164.095832  \n",
       "3     0.000000     0.000000     0.000000  \n",
       "4     0.000000     0.000000     0.000000  \n",
       "5     0.000000     0.000000     0.000000  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fc_count2fpkm(df,samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
