# AUTOGENERATED BY NBDEV! DO NOT EDIT!

__all__ = ["index", "modules", "custom_doc_links", "git_url"]

index = {"reduce_dimensional": "00_utils_rd.ipynb",
         "de_dim_methods": "00_utils_rd.ipynb",
         "de_dimensions": "01_pipelines.ipynb",
         "scatter_plot": "01_pipelines.ipynb",
         "run_main": "02_mains.ipynb",
         "scatter_plots_for_reduce_dimensional": "03_utils_plots.ipynb",
         "clusters_annotation": "04_utils_clusters.ipynb",
         "ass_methods": "04_utils_clusters.ipynb",
         "cluster_methods": "04_utils_clusters.ipynb",
         "find_peak_valley": "04_utils_clusters.ipynb",
         "find_best_cluster_number": "04_utils_clusters.ipynb",
         "cluster_params": "05_default.ipynb",
         "fc_count2tpm": "06_preprocess.ipynb",
         "fc_count2fpkm": "06_preprocess.ipynb",
         "featurecount_rename": "06_preprocess.ipynb",
         "merge_fc_deseq2": "07_merge_degfc.ipynb",
         "polya_in_nanopolish": "08_polya_sites.ipynb",
         "polya_length_in_gene": "08_polya_sites.ipynb",
         "polya_length_correlation": "08_polya_sites.ipynb",
         "parallel_run": "09_utils_multirun.ipynb",
         "filter_bam_by_name": "10_utils_bamfilter.ipynb",
         "compare_plots": "11_plots.ipynb",
         "retreat_upstreams_downstreams_from_read": "12_reads_head.ipynb",
         "retreat_seq_from_fasta": "12_reads_head.ipynb",
         "extract_sequences": "12_reads_head.ipynb"}

modules = ["utils/dedimensions.py",
           "pipelines.py",
           "run_all.py",
           "utils/plots.py",
           "utils/clusters.py",
           "defaults.py",
           "preprocess.py",
           "merge/degfc.py",
           "nanopore.py",
           "utils/multirun.py",
           "utils/bamfilter.py",
           "plots.py",
           "bam/reads.py"]

doc_url = "https://btrspg.github.io/simplebitk/"

git_url = "https://github.com/btrspg/simplebitk/tree/master/"

def custom_doc_links(name): return None
